File: node/examples/unified-streaming.ts
import { RPCClass, RPCNested } from '../src/decorators.js';
import { createRPCClient } from '../src/index.js';

@RPCClass
class TestService {
  async normalMethod(name: string): Promise<string> {
    return `Hello, ${name}!`;
  }

  async *generateNumbersService(count: number): AsyncGenerator<number> {
    for (let i = 1; i <= count; i++) {
      yield i;
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
  }

  @RPCNested
  nested = {
    async *generateData(prefix: string): AsyncGenerator<string> {
      for (let i = 1; i <= 3; i++) {
        yield `${prefix}-${i}`;
        await new Promise((resolve) => setTimeout(resolve, 50));
      }
    },
  };
}

@RPCClass
class TestHandler {
  async normalMethod(name: string): Promise<string> {
    return `RPC says: Hello, ${name}!`;
  }

  async *generateNumbers(count: number): AsyncGenerator<number> {
    for (let i = 1; i <= count; i++) {
      yield i * 10;
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
  }
}

async function main() {
  const totalStart = performance.now();
  console.log('Testing unified streaming implementation...\n');

  const server = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'unified-test-server',
  });
  await server.connect();

  const service1 = await server.service.registerHandler(
    {
      name: 'test-service',
      version: '1.0.0',
    },
    new TestService(),
  );

  const unsub1 = await server.registerHandler('test-rpc', new TestHandler());

  console.log('Server ready with service and RPC handler\n');

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'unified-test-client',
  });
  await client.connect();

  console.log('Testing Service Streaming');
  const proxyStart = performance.now();
  const service = await client.createServiceProxy<{
    normalMethod(name: string): Promise<string>;
    generateNumbersService(count: number): AsyncGenerator<number>;
    nested: {
      generateData(prefix: string): AsyncGenerator<string>;
    };
  }>('test-service');
  const proxyTime = performance.now() - proxyStart;
  console.log(`Service proxy created in ${proxyTime.toFixed(1)}ms`);

  const start = performance.now();
  const greeting = await service.normalMethod('Service');
  const methodTime = performance.now() - start;
  console.log(`Service normal method: ${greeting} (${methodTime.toFixed(1)}ms)`);

  console.log('\nService streaming:');
  let streamCount = 0;
  const streamStart = performance.now();
  for await (const num of service.generateNumbersService(5)) {
    console.log('  Received:', num);
    streamCount++;
  }
  const streamTime = performance.now() - streamStart;
  console.log(`  Total items: ${streamCount} in ${streamTime.toFixed(1)}ms`);

  console.log('\nService nested streaming:');
  let nestedCount = 0;
  const nestedStart = performance.now();
  for await (const data of service.nested.generateData('test')) {
    console.log('  Received:', data);
    nestedCount++;
  }
  const nestedTime = performance.now() - nestedStart;
  console.log(`  Total items: ${nestedCount} in ${nestedTime.toFixed(1)}ms`);

  console.log('\nTesting RPC Streaming');
  const rpcProxyStart = performance.now();
  const rpc = client.createProxy<{
    normalMethod(name: string): Promise<string>;
    generateNumbers(count: number): AsyncGenerator<number>;
  }>('test-rpc');
  const rpcProxyTime = performance.now() - rpcProxyStart;
  console.log(`RPC proxy created in ${rpcProxyTime.toFixed(1)}ms`);

  const rpcStart = performance.now();
  const rpcGreeting = await rpc.normalMethod('RPC');
  const rpcMethodTime = performance.now() - rpcStart;
  console.log(`RPC normal method: ${rpcGreeting} (${rpcMethodTime.toFixed(1)}ms)`);

  console.log('\nRPC streaming:');
  let rpcCount = 0;
  const rpcStreamStart = performance.now();
  for await (const num of rpc.generateNumbers(5)) {
    console.log('  Received:', num);
    rpcCount++;
  }
  const rpcStreamTime = performance.now() - rpcStreamStart;
  console.log(`  Total items: ${rpcCount} in ${rpcStreamTime.toFixed(1)}ms`);

  console.log('\nMessage Format Verification');
  console.log('Both service and RPC streaming now use:');
  console.log('- Same StreamMessage format');
  console.log('- Same cancellation support');
  console.log('- Same error handling');
  console.log('- Automatic chunking via client.publish()');

  console.log('\nCleaning up...');
  await service1.stop();
  await unsub1();
  await client.disconnect();
  await server.disconnect();

  const totalElapsed = performance.now() - totalStart;
  console.log(`\nTest completed successfully! Total time: ${totalElapsed.toFixed(1)}ms`);
}

main().catch(console.error);

File: python/examples/unified_streaming.py
import asyncio
import time
from collections.abc import AsyncGenerator

import uvloop

from camera_ui_rpc import RPCClass, RPCNested, ServiceConfig, create_rpc_client


@RPCClass
class TestService:
    async def normal_method(self, name: str) -> str:
        return f"Hello, {name}!"

    async def generate_numbers_service(self, count: int) -> AsyncGenerator[int, None]:
        for i in range(1, count + 1):
            yield i
            await asyncio.sleep(0.1)

    @property
    @RPCNested
    def nested(self) -> "NestedService":
        return NestedService()


@RPCClass
class NestedService:
    async def generate_data(self, prefix: str) -> AsyncGenerator[str, None]:
        for i in range(1, 4):
            yield f"{prefix}-{i}"
            await asyncio.sleep(0.05)


@RPCClass
class TestHandler:
    async def normal_method(self, name: str) -> str:
        return f"RPC says: Hello, {name}!"

    async def generate_numbers(self, count: int) -> AsyncGenerator[int, None]:
        for i in range(1, count + 1):
            yield i * 10
            await asyncio.sleep(0.1)


async def main():
    total_start = time.perf_counter()
    print("Testing unified streaming implementation...\n")

    server = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "unified-test-server",
        }
    )
    await server.connect()

    test_service = await server.service.register_handler(
        ServiceConfig(name="test-service", version="1.0.0"),
        TestService(),
    )

    unsub_rpc = await server.register_handler("test-rpc", TestHandler())

    print("Server ready with service and RPC handler\n")

    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "unified-test-client",
        }
    )
    await client.connect()

    print("=== Testing Service Streaming ===")
    proxy_start = time.perf_counter()
    service = await client.create_service_proxy("test-service", TestService)
    proxy_time = (time.perf_counter() - proxy_start) * 1000
    print(f"Service proxy created in {proxy_time:.1f}ms")

    start = time.perf_counter()
    greeting = await service.normal_method("Service")
    method_time = (time.perf_counter() - start) * 1000
    print(f"Service normal method: {greeting} ({method_time:.1f}ms)")

    print("\nService streaming:")
    stream_count = 0
    stream_start = time.perf_counter()
    async for num in service.generate_numbers_service(5):
        print("  Received:", num)
        stream_count += 1
    stream_time = (time.perf_counter() - stream_start) * 1000
    print(f"  Total items: {stream_count} in {stream_time:.1f}ms")

    print("\nService nested streaming:")
    nested_count = 0
    nested_start = time.perf_counter()
    async for data in service.nested.generate_data("test"):
        print("  Received:", data)
        nested_count += 1
    nested_time = (time.perf_counter() - nested_start) * 1000
    print(f"  Total items: {nested_count} in {nested_time:.1f}ms")

    print("\n=== Testing RPC Streaming ===")
    rpc_proxy_start = time.perf_counter()
    rpc = client.create_proxy("test-rpc", TestHandler)
    rpc_proxy_time = (time.perf_counter() - rpc_proxy_start) * 1000
    print(f"RPC proxy created in {rpc_proxy_time:.1f}ms")

    rpc_start = time.perf_counter()
    rpc_greeting = await rpc.normal_method("RPC")
    rpc_method_time = (time.perf_counter() - rpc_start) * 1000
    print(f"RPC normal method: {rpc_greeting} ({rpc_method_time:.1f}ms)")

    print("\nRPC streaming:")
    rpc_count = 0
    rpc_stream_start = time.perf_counter()
    async for num in rpc.generate_numbers(5):
        print("  Received:", num)
        rpc_count += 1
    rpc_stream_time = (time.perf_counter() - rpc_stream_start) * 1000
    print(f"  Total items: {rpc_count} in {rpc_stream_time:.1f}ms")

    print("\n=== Message Format Verification ===")
    print("Both service and RPC streaming now use:")
    print("- Same StreamMessage format")
    print("- Same cancellation support")
    print("- Same error handling")
    print("- Automatic chunking via client.publish()")

    print("\nCleaning up...")
    await test_service.stop()
    await unsub_rpc()
    await client.disconnect()
    await server.disconnect()

    total_elapsed = (time.perf_counter() - total_start) * 1000
    print(f"\nTest completed successfully! Total time: {total_elapsed:.1f}ms")


if __name__ == "__main__":
    try:
        uvloop.run(main())
    except KeyboardInterrupt:
        print("\nTest interrupted by user")
    except Exception as e:
        print(f"Error in native channel request test: {e}")
        import traceback

        traceback.print_exc()
